source/common/upstream/cluster_discovery_manager.h

```diff
@@ -0,0 +1,176 @@
+#pragma once
+
+#include <list>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "envoy/upstream/cluster_manager.h"
+
+#include "source/common/common/logger.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Upstream {
+
+/**
+ * A base class for cluster lifecycle handler. Mostly to avoid a dependency on
+ * ThreadLocalClusterManagerImpl in ClusterDiscoveryManager.
+ */
+class ClusterLifecycleCallbackHandler {
+public:
+  virtual ~ClusterLifecycleCallbackHandler() = default;
+
+  virtual ClusterUpdateCallbacksHandlePtr
+  addClusterUpdateCallbacks(ClusterUpdateCallbacks& cb) PURE;
+};
+
+/**
+ * A thread-local on-demand cluster discovery manager. It takes care of invoking the discovery
+ * callbacks in the event of a finished discovery. It does it by installing a cluster lifecycle
+ * callback that invokes the discovery callbacks when a matching cluster just got added.
+ *
+ * The manager is the sole owner of the added discovery callbacks. The only way to remove the
+ * callback from the manager is by destroying the discovery handle.
+ */
+class ClusterDiscoveryManager : Logger::Loggable<Logger::Id::upstream> {
+private:
+  struct CallbackListItem;
+  using CallbackListItemSharedPtr = std::shared_ptr<CallbackListItem>;
+  using CallbackListItemWeakPtr = std::weak_ptr<CallbackListItem>;
+  using CallbackList = std::list<CallbackListItemSharedPtr>;
+  using CallbackListIterator = CallbackList::iterator;
+
+public:
+  /**
+   * This class is used in a case when the cluster manager in the main thread notices that it
+   * already has the requested cluster, so instead of starting the discovery process, it schedules
+   * the invocation of the callback back to the thread that made the request. Invoking the request
+   * removes it from the manager.
+   */
+  class CallbackInvoker {
+  public:
+    void invokeCallback(ClusterDiscoveryStatus cluster_status) const {
+      parent_.invokeCallbackFromItem(name_, item_weak_ptr_, cluster_status);
+    }
+
+  private:
+    friend class ClusterDiscoveryManager;
+
+    CallbackInvoker(ClusterDiscoveryManager& parent, std::string name,
```

Comment:
Rvalue reference to const is a contradictory type, since by writing rvalue reference we say "we are going to move its contents into something else", while "const" means "it's immutable". I pass the string by value, since the invoker needs its own copy anyway.